### ISSUE TITLE
Update to Pay v6.0.4

### ISSUE THREAD

**Issue by joemasilotti** [joemasilotti/railsdevs.com#726]
Currently, there is some gnarly code to work around paused subscriptions. My mental model didn't line up with Stripe's when I first allowed folks to "pause" their subscription.

I thought pausing meant you should _not_ have access to the product. Stripe thinks the opposite. (I'm not 100% sure why this is the case, but this isn't the time for that.)

Today Pay released [v6.0.0](https://github.com/pay-rails/pay/releases/tag/v6.0.0) (and v6.0.4) which consolidates all the pausing behavior. Updating to this version and reworking `Businesses::Permission` will help clean up the existing (and no longer needed) code.

A PR addressing this issue will:

- [ ] Update to Pay v6.0.4
- [ ] Remove the custom "paused" logic from `Businesses::Permission`
- [ ] Ensure that:
    - [ ] A paused subscription has access until their current, active billing period ends
    - [ ] A paused subscription loses access after their active billing period ends
    - [ ] An active subscription has access
    - [ ] An inactive subscription does not have access

**Comment by sarvaiyanidhi** [comment]
Hi @joemasilotti I would like to work on this issue.. Can you please assign it to me?

**Comment by sarvaiyanidhi** [comment]
Hi @joemasilotti I have updated Pay to v6.0.3 and also added related migration for it.

In Progress - https://github.com/joemasilotti/railsdevs.com/compare/main...sarvaiyanidhi:railsdevs.com:update-pay-6.0.3

Now I see that if we pause subscription it adds current_period_start and current_period_end value in database which I understand needs to be referred to keep access active till current_period_end date.

I was accordingly trying to clean existing code in this file https://github.com/joemasilotti/railsdevs.com/blob/main/app/models/businesses/permission.rb#L34 by removing reject pause logic.

What I understood that active_or_paused scope present in this file https://github.com/pay-rails/pay/blob/master/app/models/pay/subscription.rb will automatically handle and checks current_period_end but I don't see it been checked in the scope.

So do we need to add logic in permission file to check current_period_end or am I missing anything?

**Comment by joemasilotti** [comment]
Thanks for diving in!

It looks like those fields are being used in the Stripe subscription model, not the generic one. Ideally, with this we don't need any custom "paused" logic because it all exists in Pay.

https://github.com/pay-rails/pay/blob/master/lib/pay/stripe/subscription.rb

---

Also, let's try and not run any data changes in the migration. I think we will want to sync each subscription with Stripe via Pay after the migration is run. IIRC, there's a method for that somewhere in Pay.

Let's add that to a new rake task that we can run ad-hoc now and in the future.

**Comment by sarvaiyanidhi** [comment]
Thank you Joe for the input.

Related to data changes added in migration file I did it as per the details mentioned here https://github.com/pay-rails/pay/blob/master/UPGRADE.md .. Let me know if I should move update from migration file into task. I thought it is fine as needs to be done only once while running migration.

Apart from that I have added another rake task to sync stripe subscriptions.

Now, related to pause logic I updated code in permission file by removing pause logic and added test by setting current_period_end to two days ago assuming that it will not return subscription as active but that is not the case.

https://github.com/joemasilotti/railsdevs.com/pull/735/files#diff-b05a10ef6574ec660684552b5bbf240a26b8a612609a939dd9306f0a7524c36b

I was further checking tests for Stripe subscription model https://github.com/pay-rails/pay/blob/master/test/pay/stripe/subscription_test.rb#L236 and see that it syncs subscription before checking its status but I am not very clear how we need to use it in our logic. If possible, can you please provide some insight for me to proceed further. Thanks.

**Comment by joemasilotti** [comment]
> Related to data changes added in migration file I did it as per the details mentioned here https://github.com/pay-rails/pay/blob/master/UPGRADE.md .. Let me know if I should move update from migration file into task. I thought it is fine as needs to be done only once while running migration.

Ah, good to know! Thanks for sharing that. Let's keep the data migration in the migration file then.

> Apart from that I have added another rake task to sync stripe subscriptions.

Excellent, thank you! Please add a to-do list in the PR for deploying. Even if it is just "merge, deploy, run this rake task".

> Now, related to pause logic I updated code in permission file by removing pause logic and added test by setting current_period_end to two days ago assuming that it will not return subscription as active but that is not the case.

Interesting! This pause stuff is very tricky. I wish Stripe offered a clearer API around it!

Do we need to also add the `resumes_at: nil` part of the hash like the Pay tests? Also, let's double check that we are updating the right key/value. It looks like the Pay tests are doing something a bit more deeply nested: `pause_collection: {behavior: "void"...`.

**Comment by sarvaiyanidhi** [comment]
Yes, I will update PR with steps to run rake task.

Not sure if resumes_at and pause_collection attributes are creating problem as I checked in test by adding debugger to see if subscription is paused or not and I got true for customer.subscriptions.first.paused?

Still figuring out the behavior of getting subscriptions as active though current_end_period is past and subscription is paused.

**Comment by sarvaiyanidhi** [comment]
@joemasilotti After digging further it seems like we also need to set pause_starts_at as that will get set once current_period_end  on setting subscription to pause.

So accordingly I updated pause_starts_at in test scenario and seems to pass now as subscription pause has started so it should refuse access.

Can you also have look into code once and suggest if you think something is wrong.

**Comment by joemasilotti** [comment]
Awesome! I'll comment over in the PR.

**Comment by joemasilotti** [comment]
Closed via #735.
